### PR TITLE
#4900: New model for cable paths

### PR DESCRIPTION
**Pull request by jeremystretch**
### Closes: #4900

- Introduce a new CablePath model for recording the origin, destination, intermediate nodes, and status of each end-to-end path
- Add a ForeignKey to CablePath on each of the seven path endpoint models
- Drop the old fields for `connected_endpoint` and `connection_status`
- Cache peer termination on CableTermination instances
- Remove `trace` API endpoint from FrontPort and RearPort
- Remove `connection_status` field from nested API serializers for device components
- Add `cable_peer` field to API serializer for all CableTermination models
- Overhaul cable tracing logic with the new `trace_paths()` utility function
- Completely rewrite all cable path tests
- Add the `retrace_paths` management command (temporarily?)

### CODE REVIEW

File: netbox/dcim/api/views.py [hunk 1]
```diff
@@ -469,49 +469,43 @@ def napalm(self, request, pk):
...
 
 
-class InterfaceViewSet(CableTraceMixin, ModelViewSet):
-    queryset = Interface.objects.prefetch_related(
-        'device', '_connected_interface', '_connected_circuittermination', 'cable', 'ip_addresses', 'tags'
-    ).filter(
-        device__isnull=False
-    )
+class InterfaceViewSet(PathEndpointMixin, ModelViewSet):
+    queryset = Interface.objects.prefetch_related('device', '_path', 'cable', 'ip_addresses', 'tags')
```
Comment (glennmatthews) [inline netbox/dcim/api/views.py:497]: No longer filtering out null devices?

File: netbox/dcim/management/commands/retrace_paths.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+
+class Command(BaseCommand):
+    help = "Recalculate natural ordering values for the specified models"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'args', metavar='app_label.ModelName', nargs='*',
+            help='One or more specific models (each prefixed with its app_label) to retrace',
```
Comment (glennmatthews) [inline netbox/dcim/management/commands/retrace_paths.py:27]: If this is "one or more", should `nargs='+'` instead?

**Comment by jeremystretch** [comment]
@glennmatthews I appreciate your enthusiasm, however this is a draft PR, and not yet open for review.

**Comment by jeremystretch** [comment]
To avoid jamming any complex logic into schema migrations, I've opted to retain the `trace_paths` management command and call it within `upgrade.sh`. This will, by default, automatically create any missing CablePaths. In theory, this should be necessary only when upgrading to v2.10, however I've also included a `--force` option to force the recalculation of all CablePath objects should the need arise.

I'm still mulling options for conveying the presence of multiple cable paths through a RearPort with multiple FrontPorts mapped to it in the REST API. I've dropped the `/trace/` endpoint from pass-through ports for now.

**Pull request merged**